flask_admin/contrib/fileadmin/azure.py

```diff
@@ -160,80 +151,94 @@ def get_breadcrumbs(self, path):
         )
-        return redirect(f"{url}?{sas}")
 
     def read_file(self, path):
         path = self._ensure_blob_path(path)
-
-        blob = self._client.get_blob_to_bytes(self._container_name, path)
-        return blob.content
+        if path is None:
+            raise ValueError("No path provided")
+        blob = self._container_client.get_blob_client(path).download_blob()
+        return blob.readall()
 
     def write_file(self, path, content):
         path = self._ensure_blob_path(path)
-
-        self._client.create_blob_from_text(self._container_name, path, content)
+        if path is None:
+            raise ValueError("No path provided")
+        self._container_client.upload_blob(path, content, overwrite=True)
 
     def save_file(self, path, file_data):
         path = self._ensure_blob_path(path)
-
-        self._client.create_blob_from_stream(
-            self._container_name, path, file_data.stream
-        )
+        if path is None:
+            raise ValueError("No path provided")
+        self._container_client.upload_blob(path, file_data.stream)
 
     def delete_tree(self, directory):
         directory = self._ensure_blob_path(directory)
 
-        for blob in self._client.list_blobs(self._container_name, directory):
-            self._client.delete_blob(self._container_name, blob.name)
+        for blob in self._container_client.list_blobs(directory):
+            self._container_client.delete_blob(blob.name)
 
     def delete_file(self, file_path):
         file_path = self._ensure_blob_path(file_path)
-
-        self._client.delete_blob(self._container_name, file_path)
+        if file_path is None:
+            raise ValueError("No path provided")
+        self._container_client.delete_blob(file_path)
 
     def make_dir(self, path, directory):
         path = self._ensure_blob_path(path)
         directory = self._ensure_blob_path(directory)
-
+        if path is None or directory is None:
+            raise ValueError("No path provided")
         blob = self.separator.join([path, directory, self._fakedir])
         blob = blob.lstrip(self.separator)
-        self._client.create_blob_from_text(self._container_name, blob, "")
+        self._container_client.upload_blob(blob, b"")
 
     def _copy_blob(self, src, dst):
-        src_url = self._client.make_blob_url(self._container_name, src)
-        copy = self._client.copy_blob(self._container_name, dst, src_url)
-        while copy.status != "success":
-            sleep(self._copy_poll_interval_seconds)
-            copy = self._client.get_blob_properties(
-                self._container_name, dst
-            ).properties.copy
+        src_blob_client = self._container_client.get_blob_client(src)
+        dst_blob_client = self._container_client.get_blob_client(dst)
+        copy_result = dst_blob_client.start_copy_from_url(src_blob_client.url)
+        if copy_result.get("copy_status") == "success":
```

Comment:
Is it well known in what conditions blob storage will decide to do this synchronously vs asynchronously? Is it just based on file size, or some other information?

I'm fairly uncomfortable with the sleep(10) below in the async journey. It was previously sleep(1), which is better, but I guess I'd prefer to avoid sleeping in a request at all.

I suppose if we're keeping the rename operation as a copy+delete we probably can't just move on immediately if the copy hasn't finished, so maybe this is an OK workaround if async is only going to happen for really large files. It mightstill be nice to poll for updates more frequently than every 10 seconds though?

Reply:
Good question! I was never able to get it to go down this path, even with large files. I tested up to 20 GB, and it still didn't go down the polling path. The engineer said the Blob service decides behind the scenes when to use async vs sync, and that perhaps it's always using sync for a same-account copy. I've asked if we can find that out from the Blob service team for sure.
I'm also asking if we can decrease the sleep timeout.

Reply:
Update: The SDK engineer said sleep(1) is good, so I've changed to that.